File: swift/MagCryTests/TradingTests.swift
import XCTest
@testable import MagCry

final class TradingTests: XCTestCase {

    // MARK: - Helpers

    private let playerIDs = ["You", "Alice", "Bob", "Carol", "Dave"]

    /// Standard test state: private cards sum to 43, central=[14,13,9], T=79.
    private func makeState(phase: Phase = .open) -> GameState {
        let privateCards: [String: Int] = [
            "You": 7, "Alice": 10, "Bob": 5, "Carol": 6, "Dave": 15,
        ]
        let centralCards = [14, 13, 9]
        let state = GameState(
            playerIDs: playerIDs,
            privateCards: privateCards,
            centralCards: centralCards,
            phase: phase
        )
        state.revealedCentral = centralCards  // all revealed for scoring tests
        return state
    }

    private func makeTrade(_ buyer: String, _ seller: String, _ price: Int) -> Trade {
        Trade(buyer: buyer, seller: seller, price: price, phase: .open)
    }

    // MARK: - ValidateQuote

    func testValidSpreadReturnsQuote() throws {
        let q = try validateQuote(bid: 60, ask: 62)
        XCTAssertEqual(q.bid, 60)
        XCTAssertEqual(q.ask, 62)
    }

    func testSpreadOf3Throws() {
        XCTAssertThrowsError(try validateQuote(bid: 60, ask: 63)) { error in
            XCTAssertTrue(error is TradingError)
        }
    }

    func testSpreadOf1Throws() {
        XCTAssertThrowsError(try validateQuote(bid: 60, ask: 61))
    }

    func testSpreadOf0Throws() {
        XCTAssertThrowsError(try validateQuote(bid: 60, ask: 60))
    }

    func testNegativeBidValid() throws {
        let q = try validateQuote(bid: -4, ask: -2)
        XCTAssertEqual(q.bid, -4)
        XCTAssertEqual(q.ask, -2)
    }

    // MARK: - ExecuteTradeDirectly

    func testBasicDirectTrade() throws {
        let state = makeState()
        let trade = try executeTradeDirectly(
            state: state, buyer: "You", seller: "Alice", price: 60)
        XCTAssertEqual(trade.buyer, "You")
        XCTAssertEqual(trade.seller, "Alice")
        XCTAssertEqual(trade.price, 60)
        XCTAssertEqual(trade.phase, .open)
    }

    func testTradeRecordedInState() throws {
        let state = makeState()
        try executeTradeDirectly(state: state, buyer: "You", seller: "Alice", price: 60)
        XCTAssertEqual(state.trades.count, 1)
    }

    func testMultipleDirectTradesAccumulate() throws {
        let state = makeState()
        try executeTradeDirectly(state: state, buyer: "You", seller: "Alice", price: 60)
        try executeTradeDirectly(state: state, buyer: "Bob", seller: "Carol", price: 65)
        XCTAssertEqual(state.trades.count, 2)
    }

    func testSelfTradeThrows() {
        let state = makeState()
        XCTAssertThrowsError(
            try executeTradeDirectly(state: state, buyer: "You", seller: "You", price: 60)
        ) { error in
            guard case TradingError.selfTrade = error else {
                XCTFail("Expected selfTrade error, got \(error)")
                return
            }
        }
    }

    func testSettlePhaseThrows() {
        let state = makeState(phase: .settle)
        XCTAssertThrowsError(
            try executeTradeDirectly(state: state, buyer: "You", seller: "Alice", price: 60)
        ) { error in
            guard case TradingError.settlePhase = error else {
                XCTFail("Expected settlePhase error, got \(error)")
                return
            }
        }
    }

    func testUnknownBuyerThrows() {
        let state = makeState()
        XCTAssertThrowsError(
            try executeTradeDirectly(state: state, buyer: "Ghost", seller: "Alice", price: 60)
        ) { error in
            guard case TradingError.unknownPlayer("Ghost") = error else {
                XCTFail("Expected unknownPlayer error, got \(error)")
                return
            }
        }
    }

    func testUnknownSellerThrows() {
        let state = makeState()
        XCTAssertThrowsError(
            try executeTradeDirectly(state: state, buyer: "You", seller: "Ghost", price: 60)
        ) { error in
            guard case TradingError.unknownPlayer("Ghost") = error else {
                XCTFail("Expected unknownPlayer error, got \(error)")
                return
            }
        }
    }

    func testRecordsCorrectPhase() throws {
        let state = makeState(phase: .reveal2)
        let trade = try executeTradeDirectly(
            state: state, buyer: "You", seller: "Alice", price: 58)
        XCTAssertEqual(trade.phase, .reveal2)
    }

    func testAllNonSettlePhasesAllowTrading() throws {
        for phase in [Phase.open, .reveal1, .reveal2, .reveal3] {
            let state = makeState(phase: phase)
            let trade = try executeTradeDirectly(
                state: state, buyer: "You", seller: "Alice", price: 60)
            XCTAssertEqual(trade.phase, phase)
        }
    }

    func testPnlWorksOnDirectTrade() throws {
        let state = makeState()
        let trade = try executeTradeDirectly(
            state: state, buyer: "You", seller: "Alice", price: 60)
        let T = state.finalTotal()  // 79
        XCTAssertEqual(trade.pnlFor(playerID: "You", finalTotal: T), T - 60)
        XCTAssertEqual(trade.pnlFor(playerID: "Alice", finalTotal: T), 60 - T)
    }

    func testPnlForUninvolvedPlayerIsZero() throws {
        let state = makeState()
        let trade = try executeTradeDirectly(
            state: state, buyer: "You", seller: "Alice", price: 60)
        let T = state.finalTotal()
        XCTAssertEqual(trade.pnlFor(playerID: "Bob", finalTotal: T), 0)
    }

    // MARK: - GameState

    func testFinalTotal() {
        let state = makeState()
        // 7+10+5+6+15 + 14+13+9 = 79
        XCTAssertEqual(state.finalTotal(), 79)
    }

    func testKnownCardsForPlayer() {
        let state = makeState()
        state.revealedCentral = [14]
        let known = state.knownCards(for: "You")
        XCTAssertTrue(known.contains(7))     // private card
        XCTAssertTrue(known.contains(14))    // revealed central
        XCTAssertEqual(known.count, 2)
    }

    func testKnownCardsWithNoReveals() {
        let state = makeState()
        state.revealedCentral = []
        let known = state.knownCards(for: "You")
        XCTAssertEqual(known, [7])  // just private card
    }

    func testUnknownCount() {
        let state = makeState()
        state.revealedCentral = [14]
        // 4 other players + 2 hidden central = 6
        XCTAssertEqual(state.unknownCount(for: "You"), 6)
    }

    func testUnknownCountWithAllRevealed() {
        let state = makeState()
        state.revealedCentral = [14, 13, 9]
        // 4 other players + 0 hidden central = 4
        XCTAssertEqual(state.unknownCount(for: "You"), 4)
    }

    // MARK: - Quote

    func testQuoteFromMid() {
        let q = Quote.fromMid(61.0)
        XCTAssertEqual(q.ask - q.bid, 2)
    }

    func testQuoteFromMidCentersCorrectly() {
        let q = Quote.fromMid(68.0)
        // bid = Int(68.0.rounded()) - 1 = 67, ask = 69
        XCTAssertEqual(q.bid, 67)
        XCTAssertEqual(q.ask, 69)
    }

    func testQuoteEquality() {
        let q1 = Quote(bid: 60, ask: 62)
        let q2 = Quote(bid: 60, ask: 62)
        XCTAssertEqual(q1, q2)
    }

    func testQuoteInequality() {
        let q1 = Quote(bid: 60, ask: 62)
        let q2 = Quote(bid: 58, ask: 60)
        XCTAssertNotEqual(q1, q2)
    }

    // MARK: - Phase

    func testPhaseComparable() {
        XCTAssertLessThan(Phase.open, Phase.reveal1)
        XCTAssertLessThan(Phase.reveal1, Phase.reveal2)
        XCTAssertLessThan(Phase.reveal2, Phase.reveal3)
        XCTAssertLessThan(Phase.reveal3, Phase.settle)
    }

    func testPhaseLabels() {
        XCTAssertEqual(Phase.open.label, "Open Trading")
        XCTAssertEqual(Phase.settle.label, "Settlement")
    }

    // MARK: - Settle / Scoring

    func testNoTradesAllZero() {
        let state = makeState()
        let scores = settle(state: state)
        for pid in playerIDs {
            XCTAssertEqual(scores[pid], 0)
        }
    }

    func testSingleBuyCorrectPnl() {
        let state = makeState()
        state.trades.append(makeTrade("You", "Alice", 60))
        let T = state.finalTotal()  // 79
        let scores = settle(state: state)
        XCTAssertEqual(scores["You"], T - 60)    // +19
        XCTAssertEqual(scores["Alice"], 60 - T)  // -19
    }

    func testSingleSellCorrectPnl() {
        let state = makeState()
        state.trades.append(makeTrade("Alice", "You", 70))
        let T = state.finalTotal()  // 79
        let scores = settle(state: state)
        XCTAssertEqual(scores["You"], 70 - T)     // seller: -9
        XCTAssertEqual(scores["Alice"], T - 70)    // buyer: +9
    }

    func testBuyerProfitsWhenTotalAbovePrice() {
        let state = makeState()
        state.trades.append(makeTrade("You", "Alice", 50))
        let scores = settle(state: state)
        XCTAssertGreaterThan(scores["You"]!, 0)
    }

    func testSellerProfitsWhenTotalBelowPrice() {
        // You sell at 100 (well above T=79)
        let state = makeState()
        state.trades.append(makeTrade("Alice", "You", 100))
        let scores = settle(state: state)
        XCTAssertGreaterThan(scores["You"]!, 0)
    }

    func testAllPlayersInScores() {
        let state = makeState()
        let scores = settle(state: state)
        for pid in playerIDs {
            XCTAssertNotNil(scores[pid])
        }
    }

    func testMultipleTradesAccumulatedCorrectly() {
        let state = makeState()
        state.trades.append(makeTrade("You", "Alice", 60))  // You buy at 60
        state.trades.append(makeTrade("Bob", "You", 70))    // You sell at 70
        let T = state.finalTotal()  // 79
        let scores = settle(state: state)
        let expectedYou = (T - 60) + (70 - T)  // +19 - 9 = +10
        XCTAssertEqual(scores["You"], expectedYou)
    }

    // MARK: - Gary Arbitrage

    func testArbitrageProfitIs15ForAllTotals() {
        for T in -10...130 {
            let buyPnl = T - 52
            let sellPnl = 67 - T
            XCTAssertEqual(buyPnl + sellPnl, 15,
                "Arbitrage profit should be 15 for T=\(T)")
        }
    }

    func testGaryArbitrageViaSettle() {
        // Gary buys at 52 from Alice, sells at 67 to Bob
        let privateCards: [String: Int] = [
            "You": -10, "Alice": 20, "Bob": 8, "Carol": 7, "Dave": 6,
        ]
        let centralCards = [5, 4, 3]
        // T = -10+20+8+7+6 + 5+4+3 = 43
        let state = GameState(
            playerIDs: playerIDs,
            privateCards: privateCards,
            centralCards: centralCards
        )
        state.revealedCentral = centralCards
        state.trades = [
            Trade(buyer: "You", seller: "Alice", price: 52, phase: .open),
            Trade(buyer: "Bob", seller: "You", price: 67, phase: .open),
        ]
        let scores = settle(state: state)
        XCTAssertEqual(scores["You"], 15,
            "Gary's arbitrage should profit exactly 15 (T=\(state.finalTotal()))")
    }

    func testArbitrageIsRiskFreeAcrossMultipleTotals() {
        let centralScenarios = [
            [1, 2, 3],
            [10, 11, 12],
            [13, 14, 15],
            [20, 9, 8],
        ]
        for central in centralScenarios {
            let privateCards: [String: Int] = [
                "You": -10, "Alice": 5, "Bob": 6, "Carol": 7, "Dave": 8,
            ]
            let state = GameState(
                playerIDs: playerIDs,
                privateCards: privateCards,
                centralCards: central
            )
            state.revealedCentral = central
            state.trades = [
                Trade(buyer: "You", seller: "Alice", price: 52, phase: .open),
                Trade(buyer: "Bob", seller: "You", price: 67, phase: .open),
            ]
            let scores = settle(state: state)
            XCTAssertEqual(scores["You"], 15,
                "Arbitrage failed for central=\(central), T=\(state.finalTotal())")
        }
    }

    // MARK: - Zero Sum

    func testSingleTradeZeroSum() {
        let state = makeState()
        state.trades.append(makeTrade("You", "Alice", 65))
        let scores = settle(state: state)
        XCTAssertEqual(scores.values.reduce(0, +), 0)
    }

    func testMultipleTradesZeroSum() {
        let state = makeState()
        state.trades = [
            makeTrade("You", "Alice", 58),
            makeTrade("Bob", "Carol", 65),
            makeTrade("Dave", "You", 72),
            makeTrade("Alice", "Bob", 61),
        ]
        let scores = settle(state: state)
        XCTAssertEqual(scores.values.reduce(0, +), 0)
    }

    func testManyTradesZeroSum() {
        let pairs: [(String, String, Int)] = [
            ("You", "Alice", 60), ("Bob", "Carol", 65), ("Dave", "You", 70),
            ("Alice", "Bob", 55), ("Carol", "Dave", 68), ("You", "Bob", 62),
            ("Alice", "Dave", 71), ("Carol", "You", 59), ("Bob", "Dave", 66),
            ("Dave", "Alice", 63),
        ]
        let state = makeState()
        state.trades = pairs.map { makeTrade($0.0, $0.1, $0.2) }
        let scores = settle(state: state)
        XCTAssertEqual(scores.values.reduce(0, +), 0)
    }

    // MARK: - Leaderboard

    func testLeaderboardSortedHighestFirst() {
        let scores = ["You": 10, "Alice": -5, "Bob": 25, "Carol": 0, "Dave": -15]
        let board = leaderboard(scores: scores)
        XCTAssertEqual(board[0].0, "Bob")
        XCTAssertEqual(board[0].1, 25)
        XCTAssertEqual(board.last!.0, "Dave")
    }

    func testLeaderboardAllZero() {
        let scores = ["You": 0, "Alice": 0, "Bob": 0]
        let board = leaderboard(scores: scores)
        XCTAssertEqual(board.count, 3)
        XCTAssertTrue(board.allSatisfy { $0.1 == 0 })
    }

    func testLeaderboardAllNegative() {
        let scores = ["You": -10, "Alice": -5, "Bob": -20]
        let board = leaderboard(scores: scores)
        XCTAssertEqual(board[0].0, "Alice")  // least negative = winner
    }

    func testLeaderboardSinglePlayer() {
        let scores = ["You": 42]
        let board = leaderboard(scores: scores)
        XCTAssertEqual(board[0].0, "You")
        XCTAssertEqual(board[0].1, 42)
    }

    // MARK: - Trade Breakdown

    func testTradeBreakdownFiltersToPlayerTradesOnly() {
        let state = makeState()
        state.trades = [
            makeTrade("You", "Alice", 60),
            makeTrade("Bob", "Carol", 65),  // doesn't involve You
            makeTrade("Dave", "You", 70),
        ]
        let breakdown = tradeBreakdown(state: state, playerID: "You")
        XCTAssertEqual(breakdown.count, 2)
    }

    func testUninvolvedPlayerGetsEmptyBreakdown() {
        let state = makeState()
        state.trades.append(makeTrade("Alice", "Bob", 60))
        let breakdown = tradeBreakdown(state: state, playerID: "Carol")
        XCTAssertEqual(breakdown.count, 0)
    }

    func testPnlInBreakdownIsCorrect() {
        // You buy at 60, T=79 → pnl = +19
        let state = makeState()
        state.trades.append(makeTrade("You", "Alice", 60))
        let breakdown = tradeBreakdown(state: state, playerID: "You")
        XCTAssertEqual(breakdown.count, 1)
        XCTAssertEqual(breakdown[0].pnl, state.finalTotal() - 60)
    }

    func testBreakdownIncludesBothBuyAndSell() {
        let state = makeState()
        state.trades = [
            makeTrade("You", "Alice", 60),  // You buy
            makeTrade("Bob", "You", 70),    // You sell
        ]
        let breakdown = tradeBreakdown(state: state, playerID: "You")
        XCTAssertEqual(breakdown.count, 2)
    }

    // MARK: - Difficulty

    func testDifficultyCaseIterable() {
        XCTAssertEqual(Difficulty.allCases.count, 3)
    }

    func testDifficultyRawValues() {
        XCTAssertEqual(Difficulty.easy.rawValue, 1)
        XCTAssertEqual(Difficulty.medium.rawValue, 2)
        XCTAssertEqual(Difficulty.hard.rawValue, 3)
    }

    // MARK: - Trade Identity

    func testTradeEqualityIgnoresID() {
        let t1 = Trade(buyer: "You", seller: "Alice", price: 60, phase: .open)
        let t2 = Trade(buyer: "You", seller: "Alice", price: 60, phase: .open)
        XCTAssertEqual(t1, t2)  // Equal by content, despite different UUIDs
    }

    func testTradeInequalityOnPrice() {
        let t1 = Trade(buyer: "You", seller: "Alice", price: 60, phase: .open)
        let t2 = Trade(buyer: "You", seller: "Alice", price: 65, phase: .open)
        XCTAssertNotEqual(t1, t2)
    }

    func testTradeInequalityOnPhase() {
        let t1 = Trade(buyer: "You", seller: "Alice", price: 60, phase: .open)
        let t2 = Trade(buyer: "You", seller: "Alice", price: 60, phase: .reveal1)
        XCTAssertNotEqual(t1, t2)
    }
}

File: kotlin/app/src/test/java/com/magcry/TradingTests.kt
package com.magcry

import com.magcry.model.*
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class TradingTests {

    private fun makeState(): GameState {
        val playerIDs = listOf("You", "Alice", "Bob", "Carol", "Dave")
        val privateCards = mapOf("You" to 7, "Alice" to 10, "Bob" to 5, "Carol" to 6, "Dave" to 15)
        val centralCards = listOf(14, 13, 9)
        val state = GameState(
            playerIDs = playerIDs,
            privateCards = privateCards,
            centralCards = centralCards
        )
        state.revealedCentral = centralCards.toMutableList()
        return state
    }

    // ── ValidateQuote ──

    @Test
    fun testValidSpreadReturnsQuote() {
        val q = validateQuote(58, 60)
        assertEquals(58, q.bid)
        assertEquals(60, q.ask)
    }

    @Test
    fun testSpreadOf3Throws() {
        assertThrows(TradingException.InvalidSpread::class.java) { validateQuote(58, 61) }
    }

    @Test
    fun testSpreadOf1Throws() {
        assertThrows(TradingException.InvalidSpread::class.java) { validateQuote(58, 59) }
    }

    @Test
    fun testSpreadOf0Throws() {
        assertThrows(TradingException.InvalidSpread::class.java) { validateQuote(58, 58) }
    }

    @Test
    fun testNegativeBidValid() {
        val q = validateQuote(-5, -3)
        assertEquals(-5, q.bid)
        assertEquals(-3, q.ask)
    }

    // ── ExecuteTradeDirectly ──

    @Test
    fun testBasicDirectTrade() {
        val state = makeState()
        val trade = executeTradeDirectly(state, "You", "Alice", 60)
        assertEquals("You", trade.buyer)
        assertEquals("Alice", trade.seller)
        assertEquals(60, trade.price)
    }

    @Test
    fun testTradeRecordedInState() {
        val state = makeState()
        executeTradeDirectly(state, "You", "Alice", 60)
        assertEquals(1, state.trades.size)
    }

    @Test
    fun testMultipleDirectTradesAccumulate() {
        val state = makeState()
        executeTradeDirectly(state, "You", "Alice", 60)
        executeTradeDirectly(state, "Bob", "Carol", 70)
        assertEquals(2, state.trades.size)
    }

    @Test
    fun testSelfTradeThrows() {
        val state = makeState()
        assertThrows(TradingException.SelfTrade::class.java) {
            executeTradeDirectly(state, "You", "You", 60)
        }
    }

    @Test
    fun testSettlePhaseThrows() {
        val state = makeState()
        state.phase = Phase.SETTLE
        assertThrows(TradingException.SettlePhase::class.java) {
            executeTradeDirectly(state, "You", "Alice", 60)
        }
    }

    @Test
    fun testUnknownBuyerThrows() {
        val state = makeState()
        assertThrows(TradingException.UnknownPlayer::class.java) {
            executeTradeDirectly(state, "Ghost", "Alice", 60)
        }
    }

    @Test
    fun testUnknownSellerThrows() {
        val state = makeState()
        assertThrows(TradingException.UnknownPlayer::class.java) {
            executeTradeDirectly(state, "You", "Ghost", 60)
        }
    }

    @Test
    fun testRecordsCorrectPhase() {
        val state = makeState()
        state.phase = Phase.REVEAL2
        val trade = executeTradeDirectly(state, "You", "Alice", 60)
        assertEquals(Phase.REVEAL2, trade.phase)
    }

    @Test
    fun testAllNonSettlePhasesAllowTrading() {
        val phases = listOf(Phase.OPEN, Phase.REVEAL1, Phase.REVEAL2, Phase.REVEAL3)
        for (phase in phases) {
            val state = makeState()
            state.phase = phase
            val trade = executeTradeDirectly(state, "You", "Alice", 60)
            assertEquals(phase, trade.phase)
        }
    }

    @Test
    fun testPnlWorksOnDirectTrade() {
        val state = makeState()
        val trade = executeTradeDirectly(state, "You", "Alice", 60)
        val total = state.finalTotal() // 79
        assertEquals(total - 60, trade.pnlFor("You", total))
        assertEquals(60 - total, trade.pnlFor("Alice", total))
    }

    @Test
    fun testPnlForUninvolvedPlayerIsZero() {
        val state = makeState()
        val trade = executeTradeDirectly(state, "You", "Alice", 60)
        assertEquals(0, trade.pnlFor("Bob", state.finalTotal()))
    }

    // ── GameState ──

    @Test
    fun testFinalTotal() {
        val state = makeState()
        assertEquals(79, state.finalTotal())
    }

    @Test
    fun testKnownCardsForPlayer() {
        val state = makeState()
        val known = state.knownCards("You")
        assertTrue(known.contains(7))
        assertTrue(known.contains(14))
        assertTrue(known.contains(13))
        assertTrue(known.contains(9))
    }

    @Test
    fun testKnownCardsWithNoReveals() {
        val state = makeState()
        state.revealedCentral = mutableListOf()
        val known = state.knownCards("You")
        assertEquals(1, known.size)
        assertEquals(7, known[0])
    }

    @Test
    fun testUnknownCount() {
        val state = makeState()
        // All central revealed: unknown = 4 other players
        assertEquals(4, state.unknownCount("You"))
    }

    @Test
    fun testUnknownCountWithAllRevealed() {
        val state = makeState()
        state.revealedCentral = state.centralCards.toMutableList()
        // 4 other players + 0 hidden central = 4
        assertEquals(4, state.unknownCount("You"))
    }

    // ── Quote ──

    @Test
    fun testQuoteFromMid() {
        val q = Quote.fromMid(60.0)
        assertEquals(59, q.bid)
        assertEquals(61, q.ask)
    }

    @Test
    fun testQuoteFromMidCentersCorrectly() {
        val q = Quote.fromMid(60.5)
        // round(60.5) = 61 in Kotlin (Math.round), bid = 60, ask = 62
        assertEquals(60, q.bid)
        assertEquals(62, q.ask)
    }

    @Test
    fun testQuoteEquality() {
        val a = Quote(58, 60)
        val b = Quote(58, 60)
        assertEquals(a, b)
    }

    @Test
    fun testQuoteInequality() {
        val a = Quote(58, 60)
        val b = Quote(59, 61)
        assertNotEquals(a, b)
    }

    // ── Phase ──

    @Test
    fun testPhaseComparable() {
        assertTrue(Phase.OPEN.ordinal < Phase.REVEAL1.ordinal)
        assertTrue(Phase.REVEAL1.ordinal < Phase.REVEAL2.ordinal)
        assertTrue(Phase.REVEAL2.ordinal < Phase.REVEAL3.ordinal)
        assertTrue(Phase.REVEAL3.ordinal < Phase.SETTLE.ordinal)
    }

    @Test
    fun testPhaseLabels() {
        assertEquals("Open Trading", Phase.OPEN.label)
        assertEquals("Reveal 1", Phase.REVEAL1.label)
        assertEquals("Reveal 2", Phase.REVEAL2.label)
        assertEquals("Reveal 3", Phase.REVEAL3.label)
        assertEquals("Settlement", Phase.SETTLE.label)
    }

    // ── Settle / Scoring ──

    @Test
    fun testNoTradesAllZero() {
        val state = makeState()
        val scores = settle(state)
        for ((_, pnl) in scores) {
            assertEquals(0, pnl)
        }
    }

    @Test
    fun testSingleBuyCorrectPnl() {
        val state = makeState()
        executeTradeDirectly(state, "You", "Alice", 60)
        val scores = settle(state)
        assertEquals(79 - 60, scores["You"])
        assertEquals(60 - 79, scores["Alice"])
    }

    @Test
    fun testSingleSellCorrectPnl() {
        val state = makeState()
        executeTradeDirectly(state, "Alice", "You", 60)
        val scores = settle(state)
        assertEquals(60 - 79, scores["You"])
        assertEquals(79 - 60, scores["Alice"])
    }

    @Test
    fun testBuyerProfitsWhenTotalAbovePrice() {
        val state = makeState() // total = 79
        executeTradeDirectly(state, "You", "Alice", 70)
        val scores = settle(state)
        assertTrue(scores["You"]!! > 0)
    }

    @Test
    fun testSellerProfitsWhenTotalBelowPrice() {
        val state = makeState() // total = 79
        executeTradeDirectly(state, "Alice", "You", 85)
        val scores = settle(state)
        assertTrue(scores["You"]!! > 0) // You sold at 85, total 79 => +6
    }

    @Test
    fun testAllPlayersInScores() {
        val state = makeState()
        val scores = settle(state)
        for (pid in state.playerIDs) {
            assertTrue(scores.containsKey(pid))
        }
    }

    @Test
    fun testMultipleTradesAccumulatedCorrectly() {
        val state = makeState()
        executeTradeDirectly(state, "You", "Alice", 60)
        executeTradeDirectly(state, "You", "Bob", 70)
        val scores = settle(state)
        // You bought twice: (79-60) + (79-70) = 19+9 = 28
        assertEquals(28, scores["You"])
    }

    // ── Arbitrage ──

    @Test
    fun testArbitrageProfitIs15ForAllTotals() {
        // Buy at 60, sell at 75 => P&L = (T-60) + (75-T) = 15 regardless of T
        val state = makeState()
        executeTradeDirectly(state, "You", "Alice", 60)
        executeTradeDirectly(state, "Bob", "You", 75)
        val scores = settle(state)
        assertEquals(15, scores["You"])
    }

    @Test
    fun testGaryArbitrageViaSettle() {
        // "Gary" is "You" — same arb test via settle
        val state = makeState()
        executeTradeDirectly(state, "You", "Alice", 60)
        executeTradeDirectly(state, "Bob", "You", 75)
        val scores = settle(state)
        assertEquals(15, scores["You"])
    }

    @Test
    fun testArbitrageIsRiskFreeAcrossMultipleTotals() {
        // Test with different card setups; arb always = ask - bid
        for (total in listOf(50, 79, 100, 130)) {
            val buyPrice = 60
            val sellPrice = 75
            val arbProfit = sellPrice - buyPrice
            assertEquals(15, arbProfit)
        }
    }

    // ── Zero Sum ──

    @Test
    fun testSingleTradeZeroSum() {
        val state = makeState()
        executeTradeDirectly(state, "You", "Alice", 60)
        val scores = settle(state)
        assertEquals(0, scores.values.sum())
    }

    @Test
    fun testMultipleTradesZeroSum() {
        val state = makeState()
        executeTradeDirectly(state, "You", "Alice", 60)
        executeTradeDirectly(state, "Bob", "Carol", 70)
        val scores = settle(state)
        assertEquals(0, scores.values.sum())
    }

    @Test
    fun testManyTradesZeroSum() {
        val state = makeState()
        executeTradeDirectly(state, "You", "Alice", 60)
        executeTradeDirectly(state, "Bob", "Carol", 70)
        executeTradeDirectly(state, "Dave", "You", 80)
        executeTradeDirectly(state, "Alice", "Bob", 65)
        val scores = settle(state)
        assertEquals(0, scores.values.sum())
    }

    // ── Leaderboard ──

    @Test
    fun testLeaderboardSortedHighestFirst() {
        val state = makeState()
        executeTradeDirectly(state, "You", "Alice", 60)
        executeTradeDirectly(state, "Bob", "Carol", 70)
        val scores = settle(state)
        val lb = leaderboard(scores)
        for (i in 0 until lb.size - 1) {
            assertTrue(lb[i].second >= lb[i + 1].second)
        }
    }

    @Test
    fun testLeaderboardAllZero() {
        val state = makeState()
        val scores = settle(state)
        val lb = leaderboard(scores)
        for ((_, pnl) in lb) {
            assertEquals(0, pnl)
        }
    }

    @Test
    fun testLeaderboardAllNegative() {
        // Everyone sells to someone at a price above total (79)
        val state = makeState()
        executeTradeDirectly(state, "Alice", "You", 100)
        executeTradeDirectly(state, "Bob", "Carol", 100)
        val scores = settle(state)
        val lb = leaderboard(scores)
        // Sorted highest first
        assertTrue(lb[0].second >= lb[1].second)
    }

    @Test
    fun testLeaderboardSinglePlayer() {
        val scores = mapOf("You" to 10)
        val lb = leaderboard(scores)
        assertEquals(1, lb.size)
        assertEquals("You", lb[0].first)
        assertEquals(10, lb[0].second)
    }

    // ── TradeBreakdown ──

    @Test
    fun testTradeBreakdownFiltersToPlayerTradesOnly() {
        val state = makeState()
        executeTradeDirectly(state, "You", "Alice", 60)
        executeTradeDirectly(state, "Bob", "Carol", 70)
        val bd = tradeBreakdown(state, "You")
        assertEquals(1, bd.size)
    }

    @Test
    fun testUninvolvedPlayerGetsEmptyBreakdown() {
        val state = makeState()
        executeTradeDirectly(state, "You", "Alice", 60)
        val bd = tradeBreakdown(state, "Dave")
        assertEquals(0, bd.size)
    }

    @Test
    fun testPnlInBreakdownIsCorrect() {
        val state = makeState()
        executeTradeDirectly(state, "You", "Alice", 60)
        val bd = tradeBreakdown(state, "You")
        assertEquals(79 - 60, bd[0].second)
    }

    @Test
    fun testBreakdownIncludesBothBuyAndSell() {
        val state = makeState()
        executeTradeDirectly(state, "You", "Alice", 60)
        executeTradeDirectly(state, "Bob", "You", 70)
        val bd = tradeBreakdown(state, "You")
        assertEquals(2, bd.size)
    }

    // ── Difficulty ──

    @Test
    fun testDifficultyCaseIterable() {
        assertEquals(3, Difficulty.entries.size)
    }

    @Test
    fun testDifficultyRawValues() {
        assertEquals(1, Difficulty.EASY.value)
        assertEquals(2, Difficulty.MEDIUM.value)
        assertEquals(3, Difficulty.HARD.value)
    }

    // ── Trade Identity ──

    @Test
    fun testTradeEqualityIgnoresID() {
        val a = Trade(buyer = "You", seller = "Alice", price = 60, phase = Phase.OPEN)
        val b = Trade(buyer = "You", seller = "Alice", price = 60, phase = Phase.OPEN)
        assertEquals(a, b)
    }

    @Test
    fun testTradeInequalityOnPrice() {
        val a = Trade(buyer = "You", seller = "Alice", price = 60, phase = Phase.OPEN)
        val b = Trade(buyer = "You", seller = "Alice", price = 65, phase = Phase.OPEN)
        assertNotEquals(a, b)
    }

    @Test
    fun testTradeInequalityOnPhase() {
        val a = Trade(buyer = "You", seller = "Alice", price = 60, phase = Phase.OPEN)
        val b = Trade(buyer = "You", seller = "Alice", price = 60, phase = Phase.REVEAL1)
        assertNotEquals(a, b)
    }
}
